firebase-firestore/src/test/java/com/google/firebase/firestore/FilterTest.java

```diff
@@ -0,0 +1,188 @@
+    assertEquals(filter, Filter.lessThanOrEqualTo(FieldPath.of("x"), "v"));
+    assertNotEquals(filter, Filter.lessThanOrEqualTo("x", "z"));
+    assertNotEquals(filter, Filter.lessThanOrEqualTo("y", "v"));
+    assertNotEquals(filter,Filter.greaterThanOrEqualTo("y", "v"));
+  }
+
+  @Test
+  public void arrayContains() {
+    Filter filter = Filter.arrayContains("x", "v");
+    assertEquals(filter, filter);
+    assertEquals(filter, Filter.arrayContains(FieldPath.of("x"), "v"));
+    assertNotEquals(filter, Filter.arrayContains("x", "z"));
+    assertNotEquals(filter, Filter.arrayContains("y", "v"));
+    assertNotEquals(filter,Filter.equalTo("y", "v"));
+  }
+
+  @Test
+  public void arrayContainsAny() {
+    Filter filter = Filter.arrayContainsAny("x", ImmutableList.of("v1", "v2"));
+    assertEquals(filter, filter);
+    assertEquals(filter, Filter.arrayContainsAny(FieldPath.of("x"), ImmutableList.of("v1", "v2")));
+    assertNotEquals(filter, Filter.arrayContainsAny("x", ImmutableList.of("v2", "v3")));
+    assertNotEquals(filter, Filter.arrayContainsAny("y", ImmutableList.of("v1", "v2")));
+    assertNotEquals(filter, Filter.equalTo("y", "v"));
+  }
+
+  @Test
+  public void inArray() {
+    Filter filter = Filter.inArray("x", ImmutableList.of("v1", "v2"));
+    assertEquals(filter, filter);
+    assertEquals(filter, Filter.inArray(FieldPath.of("x"), ImmutableList.of("v1", "v2")));
+    assertNotEquals(filter, Filter.inArray("x", ImmutableList.of("v2", "v3")));
+    assertNotEquals(filter, Filter.inArray("y", ImmutableList.of("v1", "v2")));
+    assertNotEquals(filter, Filter.notInArray("y", ImmutableList.of("v1", "v2")));
+  }
+
+  @Test
+  public void notInArray() {
+    Filter filter = Filter.notInArray("x", ImmutableList.of("v1", "v2"));
+    assertEquals(filter, filter);
+    assertEquals(filter, Filter.notInArray(FieldPath.of("x"), ImmutableList.of("v1", "v2")));
+    assertNotEquals(filter, Filter.notInArray("x", ImmutableList.of("v2", "v3")));
+    assertNotEquals(filter, Filter.notInArray("y", ImmutableList.of("v1", "v2")));
+    assertNotEquals(filter, Filter.inArray("y", ImmutableList.of("v1", "v2")));
+  }
+
+  @Test
+  public void or() {
+    Filter filter = Filter.or(
+        Filter.equalTo("x", ImmutableList.of("v1", "v2")),
+        Filter.equalTo("y", ImmutableList.of("v3", "v4"))
+    );
+    assertEquals(
+        filter,
+        Filter.or(
+            Filter.equalTo(FieldPath.of("x"), ImmutableList.of("v1", "v2")),
+            Filter.equalTo(FieldPath.of("y"), ImmutableList.of("v3", "v4"))
+        )
+    );
+    assertNotEquals(
+        filter,
+        Filter.and(
+            Filter.equalTo("x", ImmutableList.of("v1", "v2")),
+            Filter.equalTo("y", ImmutableList.of("v3", "v4"))
+        )
+    );
+    assertNotEquals(
+        filter,
+        Filter.and(
+            Filter.equalTo("y", ImmutableList.of("v3", "v4")),
+            Filter.equalTo("x", ImmutableList.of("v1", "v2"))
+        )
+    );
```

Comment:
add a case where the operator remains unchanged, but the _filter_ order changes.

i.e.
```
assertNotEquals(
        filter,
        Filter.or(
            Filter.equalTo("y", ImmutableList.of("v3", "v4")),
            Filter.equalTo("x", ImmutableList.of("v1", "v2"))
        )
    );
```

add a case where the operator remains unchanged, and the filter order remains unchanged, but the list value order changes.

i.e.
```
assertNotEquals(
        filter,
        Filter.or(
          Filter.equalTo("x", ImmutableList.of("v2", "v1")),
          Filter.equalTo("y", ImmutableList.of("v4", "v3"))
        )
    );
```

Reply:
Right. That was the intent of last assert, so I will fix it to match what you say.

Reply:
Sorry, I don't see point of second test you want to add. That is already covered in the `inArray` tests, after I switched operator to be `inArray`.

Reply:
No wait... It should have been covered, but I actually changed the values, not just the order. I will add test up there. I think the And/Or operators need to be trusted to call equals methods on composed Filters.